File: Flutter/Exemple_flutter/lib/sdk/smartc_client.dart
// =============================================================================
// SmaRTC Flutter SDK - Client simplifié
// =============================================================================

import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:signalr_netcore/signalr_client.dart';

/// Configuration du client SmaRTC
class SmaRTCConfig {
  final String apiBaseUrl;
  final String signalHubUrl;

  const SmaRTCConfig({
    this.apiBaseUrl = 'http://localhost:8080',
    this.signalHubUrl = 'http://localhost:5001/signalhub',
  });
}

/// Modèle utilisateur
class SmaRTCUser {
  final String id;
  final String username;
  final String? token;

  SmaRTCUser({required this.id, required this.username, this.token});

  factory SmaRTCUser.fromJson(Map<String, dynamic> json) {
    return SmaRTCUser(
      id: json['id']?.toString() ?? '',
      username: json['username'] ?? '',
      token: json['token'],
    );
  }
}

/// Message de chat
class ChatMessage {
  final String sender;
  final String content;
  final DateTime timestamp;
  final bool isSystem;

  ChatMessage({
    required this.sender,
    required this.content,
    required this.timestamp,
    this.isSystem = false,
  });
}

/// Client SmaRTC principal
class SmaRTCClient {
  final SmaRTCConfig config;
  HubConnection? _hubConnection;
  SmaRTCUser? _currentUser;
  String? _currentRoom;
  bool _isConnected = false;

  // Callbacks
  Function(ChatMessage)? onMessageReceived;
  Function(String)? onUserJoined;
  Function(String)? onUserLeft;
  Function(String)? onError;
  Function()? onConnected;
  Function()? onDisconnected;

  SmaRTCClient({SmaRTCConfig? config}) : config = config ?? const SmaRTCConfig();

  /// Utilisateur actuel
  SmaRTCUser? get currentUser => _currentUser;

  /// Room actuelle
  String? get currentRoom => _currentRoom;

  /// État de connexion
  bool get isConnected => _isConnected;

  /// Inscription d'un nouvel utilisateur
  Future<bool> registerAsync(String username, String password) async {
    try {
      final response = await http.post(
        Uri.parse('${config.apiBaseUrl}/api/auth/register'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'username': username,
          'password': password,
          'email': '$username@smaRTC.local',
        }),
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        return true;
      } else if (response.statusCode == 400 || response.statusCode == 409) {
        // Utilisateur existe déjà - OK pour login
        final body = response.body.toLowerCase();
        if (body.contains('already') || body.contains('exists')) {
          return true;
        }
      }
      return false;
    } catch (e) {
      onError?.call('Erreur d\'inscription: $e');
      return false;
    }
  }

  /// Connexion utilisateur
  Future<bool> loginAsync(String username, String password) async {
    try {
      final response = await http.post(
        Uri.parse('${config.apiBaseUrl}/api/auth/login'),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'username': username,
          'password': password,
        }),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        _currentUser = SmaRTCUser(
          id: data['userId']?.toString() ?? data['id']?.toString() ?? '',
          username: username,
          token: data['token'],
        );
        return true;
      }
      return false;
    } catch (e) {
      onError?.call('Erreur de connexion: $e');
      return false;
    }
  }

  /// Connexion au hub SignalR
  Future<bool> connectToHubAsync() async {
    try {
      _hubConnection = HubConnectionBuilder()
          .withUrl(config.signalHubUrl)
          .withAutomaticReconnect()
          .build();

      // Handlers pour les événements
      _hubConnection!.on('ReceiveSignal', _handleReceiveSignal);
      _hubConnection!.on('UserJoined', _handleUserJoined);
      _hubConnection!.on('UserLeft', _handleUserLeft);
      _hubConnection!.on('ReceiveMessage', _handleReceiveMessage);

      _hubConnection!.onclose(({error}) {
        _isConnected = false;
        onDisconnected?.call();
      });

      _hubConnection!.onreconnecting(({error}) {
        _isConnected = false;
      });

      _hubConnection!.onreconnected(({connectionId}) {
        _isConnected = true;
        // Rejoindre la room si on était dans une
        if (_currentRoom != null) {
          joinRoomAsync(_currentRoom!);
        }
      });

      await _hubConnection!.start();
      _isConnected = true;
      onConnected?.call();
      return true;
    } catch (e) {
      onError?.call('Erreur de connexion au hub: $e');
      return false;
    }
  }

  /// Rejoindre une room
  Future<bool> joinRoomAsync(String roomName) async {
    if (_hubConnection == null || !_isConnected) {
      onError?.call('Non connecté au hub');
      return false;
    }

    try {
      await _hubConnection!.invoke('JoinSession', args: [roomName]);
      _currentRoom = roomName;
      
      // Message système
      onMessageReceived?.call(ChatMessage(
        sender: 'Système',
        content: '${_currentUser?.username ?? "Vous"} a rejoint la room $roomName',
        timestamp: DateTime.now(),
        isSystem: true,
      ));
      
      return true;
    } catch (e) {
      onError?.call('Erreur pour rejoindre la room: $e');
      return false;
    }
  }

  /// Quitter la room
  Future<void> leaveRoomAsync() async {
    if (_hubConnection == null || _currentRoom == null) return;

    try {
      await _hubConnection!.invoke('LeaveSession', args: [_currentRoom]);
      _currentRoom = null;
    } catch (e) {
      onError?.call('Erreur pour quitter la room: $e');
    }
  }

  /// Envoyer un message
  Future<bool> sendMessageAsync(String message) async {
    if (_hubConnection == null || _currentRoom == null || !_isConnected) {
      onError?.call('Non connecté ou pas dans une room');
      return false;
    }

    try {
      final signalData = jsonEncode({
        'type': 'chat',
        'sender': _currentUser?.username ?? 'Anonymous',
        'content': message,
        'timestamp': DateTime.now().toIso8601String(),
      });

      await _hubConnection!.invoke('SendSignalToSession', args: [_currentRoom, signalData]);
      
      // Ajouter le message localement
      onMessageReceived?.call(ChatMessage(
        sender: _currentUser?.username ?? 'Moi',
        content: message,
        timestamp: DateTime.now(),
      ));
      
      return true;
    } catch (e) {
      onError?.call('Erreur d\'envoi du message: $e');
      return false;
    }
  }

  /// Déconnexion
  Future<void> disconnectAsync() async {
    if (_currentRoom != null) {
      await leaveRoomAsync();
    }
    
    if (_hubConnection != null) {
      await _hubConnection!.stop();
      _hubConnection = null;
    }
    
    _isConnected = false;
    _currentUser = null;
    onDisconnected?.call();
  }

  // ============= Handlers privés =============

  void _handleReceiveSignal(List<Object?>? args) {
    if (args == null || args.isEmpty) return;

    try {
      final signalData = args[0]?.toString() ?? '';
      final data = jsonDecode(signalData);
      
      if (data['type'] == 'chat') {
        final sender = data['sender'] ?? 'Unknown';
        // Ne pas afficher nos propres messages (déjà affichés)
        if (sender != _currentUser?.username) {
          onMessageReceived?.call(ChatMessage(
            sender: sender,
            content: data['content'] ?? '',
            timestamp: DateTime.tryParse(data['timestamp'] ?? '') ?? DateTime.now(),
          ));
        }
      }
    } catch (e) {
      // Ignorer les erreurs de parsing
    }
  }

  void _handleUserJoined(List<Object?>? args) {
    if (args != null && args.isNotEmpty) {
      final username = args[0]?.toString() ?? 'Unknown';
      onUserJoined?.call(username);
      onMessageReceived?.call(ChatMessage(
        sender: 'Système',
        content: '$username a rejoint la room',
        timestamp: DateTime.now(),
        isSystem: true,
      ));
    }
  }

  void _handleUserLeft(List<Object?>? args) {
    if (args != null && args.isNotEmpty) {
      final username = args[0]?.toString() ?? 'Unknown';
      onUserLeft?.call(username);
      onMessageReceived?.call(ChatMessage(
        sender: 'Système',
        content: '$username a quitté la room',
        timestamp: DateTime.now(),
        isSystem: true,
      ));
    }
  }

  void _handleReceiveMessage(List<Object?>? args) {
    if (args == null || args.length < 2) return;
    
    final sender = args[0]?.toString() ?? 'Unknown';
    final content = args[1]?.toString() ?? '';
    
    if (sender != _currentUser?.username) {
      onMessageReceived?.call(ChatMessage(
        sender: sender,
        content: content,
        timestamp: DateTime.now(),
      ));
    }
  }
}

File: C#/Exemple_csharp/SDK/SmaRTCClient.cs
using System.Net.Http.Headers;
using System.Net.Http.Json;
using System.Text.Json;
using System.Text.Json.Serialization;
using Microsoft.AspNetCore.SignalR.Client;

namespace Exemple_csharp.SDK
{
    /// <summary>
    /// Client SmaRTC pour la démonstration C#
    /// Gère l'authentification, les sessions et le chat en temps réel
    /// </summary>
    public class SmaRTCClient : IAsyncDisposable
    {
        private readonly HttpClient _httpClient;
        private readonly string _apiUrl;
        private readonly string _signalUrl;
        private HubConnection? _hubConnection;
        private string? _token;
        private string? _username;
        private string? _currentRoom;

        // Events
        public event Action<string, string>? OnMessageReceived;
        public event Action<string>? OnUserJoined;
        public event Action<string>? OnUserLeft;
        public event Action<string>? OnError;
        public event Action? OnConnected;
        public event Action? OnDisconnected;

        // Properties
        public bool IsAuthenticated => !string.IsNullOrEmpty(_token);
        public bool IsConnected => _hubConnection?.State == HubConnectionState.Connected;
        public string? Username => _username;
        public string? CurrentRoom => _currentRoom;

        public SmaRTCClient(string apiUrl = "http://localhost:8080", string signalUrl = "http://localhost:5001")
        {
            _apiUrl = apiUrl;
            _signalUrl = signalUrl;
            _httpClient = new HttpClient
            {
                BaseAddress = new Uri(apiUrl),
                Timeout = TimeSpan.FromSeconds(30)
            };
        }

        #region Authentication

        /// <summary>
        /// Enregistre un nouvel utilisateur
        /// Retourne true si l'inscription réussit ou si l'utilisateur existe déjà
        /// </summary>
        public async Task<bool> RegisterAsync(string username, string email, string password)
        {
            try
            {
                var request = new
                {
                    Username = username,
                    Email = email,
                    Password = password,
                    Role = "User"
                };

                var response = await _httpClient.PostAsJsonAsync("/api/Auth/register", request);
                
                if (response.IsSuccessStatusCode)
                {
                    return true;
                }
                
                var error = await response.Content.ReadAsStringAsync();
                
                // Si l'utilisateur existe déjà, ce n'est pas une vraie erreur
                if (error.Contains("already exists", StringComparison.OrdinalIgnoreCase) ||
                    error.Contains("Username already", StringComparison.OrdinalIgnoreCase))
                {
                    return true; // L'utilisateur existe déjà, on peut continuer
                }
                
                OnError?.Invoke($"Erreur d'inscription: {error}");
                return false;
            }
            catch (Exception ex)
            {
                OnError?.Invoke($"Erreur réseau: {ex.Message}");
                return false;
            }
        }

        /// <summary>
        /// Authentifie l'utilisateur
        /// </summary>
        public async Task<bool> LoginAsync(string username, string password)
        {
            try
            {
                var request = new
                {
                    Username = username,
                    Password = password
                };

                var response = await _httpClient.PostAsJsonAsync("/api/Auth/login", request);
                
                if (response.IsSuccessStatusCode)
                {
                    var result = await response.Content.ReadFromJsonAsync<LoginResponse>();
                    if (result?.Token != null)
                    {
                        _token = result.Token;
                        _username = username;
                        _httpClient.DefaultRequestHeaders.Authorization = 
                            new AuthenticationHeaderValue("Bearer", _token);
                        return true;
                    }
                }
                
                var error = await response.Content.ReadAsStringAsync();
                OnError?.Invoke($"Erreur de connexion: {error}");
                return false;
            }
            catch (Exception ex)
            {
                OnError?.Invoke($"Erreur réseau: {ex.Message}");
                return false;
            }
        }

        #endregion

        #region SignalR Connection

        /// <summary>
        /// Connecte au serveur SignalR pour le chat temps réel
        /// </summary>
        public async Task<bool> ConnectToSignalRAsync()
        {
            if (!IsAuthenticated)
            {
                OnError?.Invoke("Vous devez être authentifié avant de vous connecter au chat");
                return false;
            }

            try
            {
                _hubConnection = new HubConnectionBuilder()
                    .WithUrl($"{_signalUrl}/signalhub", options =>
                    {
                        options.AccessTokenProvider = () => Task.FromResult(_token);
                    })
                    .WithAutomaticReconnect()
                    .Build();

                // Setup event handlers - adapté aux méthodes du SignalHub
                _hubConnection.On<string>("NewUserArrived", (user) =>
                {
                    OnUserJoined?.Invoke(user);
                });

                _hubConnection.On<string>("UserLeft", (user) =>
                {
                    OnUserLeft?.Invoke(user);
                });

                // Pour le chat, on utilise SendSignal avec un format JSON
                _hubConnection.On<string, string>("SendSignal", (signal, user) =>
                {
                    // Traiter le signal comme un message
                    OnMessageReceived?.Invoke(user, signal);
                });

                _hubConnection.Closed += async (error) =>
                {
                    OnDisconnected?.Invoke();
                    if (error != null)
                    {
                        OnError?.Invoke($"Connexion perdue: {error.Message}");
                    }
                    await Task.CompletedTask;
                };

                _hubConnection.Reconnected += async (connectionId) =>
                {
                    OnConnected?.Invoke();
                    // Rejoindre automatiquement la room après reconnexion
                    if (!string.IsNullOrEmpty(_currentRoom))
                    {
                        await JoinRoomAsync(_currentRoom);
                    }
                    await Task.CompletedTask;
                };

                await _hubConnection.StartAsync();
                OnConnected?.Invoke();
                return true;
            }
            catch (Exception ex)
            {
                OnError?.Invoke($"Erreur de connexion SignalR: {ex.Message}");
                return false;
            }
        }

        #endregion

        #region Room Management

        /// <summary>
        /// Rejoint une room de chat (session SignalR)
        /// </summary>
        public async Task<bool> JoinRoomAsync(string roomName)
        {
            if (_hubConnection?.State != HubConnectionState.Connected)
            {
                OnError?.Invoke("Non connecté au serveur de chat");
                return false;
            }

            try
            {
                // Utilise JoinSession du SignalHub
                await _hubConnection.InvokeAsync("JoinSession", roomName, _username);
                _currentRoom = roomName;
                return true;
            }
            catch (Exception ex)
            {
                OnError?.Invoke($"Erreur pour rejoindre la room: {ex.Message}");
                return false;
            }
        }

        /// <summary>
        /// Quitte la room actuelle
        /// </summary>
        public async Task<bool> LeaveRoomAsync()
        {
            if (_hubConnection?.State != HubConnectionState.Connected || string.IsNullOrEmpty(_currentRoom))
            {
                return false;
            }

            try
            {
                // Utilise LeaveSession du SignalHub
                await _hubConnection.InvokeAsync("LeaveSession", _currentRoom, _username);
                _currentRoom = null;
                return true;
            }
            catch (Exception ex)
            {
                OnError?.Invoke($"Erreur pour quitter la room: {ex.Message}");
                return false;
            }
        }

        #endregion

        #region Messaging

        /// <summary>
        /// Envoie un message dans la room actuelle (utilise SendSignalToSession)
        /// </summary>
        public async Task<bool> SendMessageAsync(string message)
        {
            if (_hubConnection?.State != HubConnectionState.Connected)
            {
                OnError?.Invoke("Non connecté au serveur de chat");
                return false;
            }

            if (string.IsNullOrEmpty(_currentRoom))
            {
                OnError?.Invoke("Vous n'êtes dans aucune room");
                return false;
            }

            try
            {
                // Utilise SendSignalToSession du SignalHub pour envoyer des messages
                await _hubConnection.InvokeAsync("SendSignalToSession", _currentRoom, message, _username);
                return true;
            }
            catch (Exception ex)
            {
                OnError?.Invoke($"Erreur d'envoi: {ex.Message}");
                return false;
            }
        }

        #endregion

        #region Cleanup

        public async ValueTask DisposeAsync()
        {
            if (_hubConnection != null)
            {
                await _hubConnection.DisposeAsync();
            }
            _httpClient.Dispose();
        }

        #endregion

        #region Models

        private class LoginResponse
        {
            [JsonPropertyName("token")]
            public string? Token { get; set; }
        }

        #endregion
    }
}
